tensor_comprehensions/__init__.py

```diff
@@ -12,20 +12,372 @@
+            unchecked: Optional[bool] = False) -> List[torch.Tensor]:
+        r"""
+        Runs the defined TC function on given inputs. If the TC function has not
+        yet been compiled for the particular input sizes it will be JIT-compiled
+        and memoized the first time. The MappingOptions used to compile the TC
+        function is determined by the TC object's  mapping_options_factory which
+        can be (re)set for a TC object by calling set_mapping_options_factory.
+
+        Args:
+            entry_point: the name of the TC def to be executed
+
+            inputs: tensors that TC takes as input arguments. The inputs must
+                be passed in the order of the definition of the TC function.
+
+            outputs: tensors the TC writes to in-place. If output tensors are
+                omitted, new outputs will be allocated and returned at each
+                call.
+
+            unchecked: run in low-latency unchecked mode where failsafe size
+                and stride checks are omitted.
+
+        Returns
+            List[torch.Tensor]: the tensors returned by the TC. If outputs are
+                not specified then new allocations take place. Otherwise the TC
+                writes in-place and returns the same tensors it received as
+                outputs.
+        """
+
+        # Locally scoped implicit compilation
+        def __implicit_compile__(tc_obj: TC,
+                                 entry_point: str,
+                                 *inputs: torch.Tensor):
+            already_compiled = tc_obj.compilation_cache.is_compiled(
+                entry_point, inputs)
+
+            if already_compiled:
+                return
+
+            if not SILENT:
+                print(
+                    "TC \"{}\" was not explicitly compiled for ".format(entry_point) +
+                    "inputs of sizes:\n  {}\n".format("".join("{}, ".format(
+                        i.size().__str__()) for i in inputs)) +
+                    "....Generate implicit MappingOptions")
+
+            mapping_options = tc_obj.mapping_options_factory(
+                tc_obj.tc, entry_point, *inputs)
+
+            assert mapping_options is not None, (
+                "No options found for TC {} ".format(entry_point) +
+                "with inputs of sizes:\n  {}\n".format(
+                    "".join("{} ".format(i.size().__str__()) for i in inputs)))
+
+            # Compile best options to set the executor for the current
+            #     (entry point, inputs)
+            start = time.clock()
+            tc_obj.compilation_cache.compile(
+                entry_point, inputs, mapping_options)
+            if not SILENT:
+                print(
+                    "Done compiling TC \"{}\" (compile time: {}ms)".format(
+                        entry_point, int((time.clock() - start) * 10 ** 3)))
+
+        __implicit_compile__(self, entry_point, *inputs)
```

Comment:
Why define a function in a function scope and only call it once? I'd just paste the code and add the comments

Reply:
makes it as explicit and structured as possible that this is an implicit behavior